src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java

```diff
@@ -55,6 +55,13 @@
     boolean autoSave;
     boolean autoBuildRoleLinks;
 
+    // cached instance of AviatorEvaluatorInstance
+    AviatorEvaluatorInstance aviatorEval;
+
+    // keep track of changes in Model and FunctionMap so that we can invalidate cached AviatorEvaluatorInstance
+    int modelModCount;
```

Comment:
The model API is undocumented and should be called at one's own risk. We will not update the evaluator cache in that case.

Reply:
ok, good to know. 

Is it the same rule for FunctionsMap.addFunction? do we support only calling enforcer.addFunction?

Same for Helper.loadPolicyLine: is it an undocumented API?

What about Model.addDef? there is a javadoc on this method, so it seems that it is allowed to do this:
```
Model m = CoreEnforcer.newModel();
(populate model)

Enforcer e = new Enforcer(m);
e.enforce(...); // cache will be built here

m.addDef(...); // cache is now invalid, how do I detect this?

```

Reply:
I think `Model.addDef()` can be a public API. So Model can have a pointer to Enforcer. But `FunctionsMap.addFunction()` and `Helper.loadPolicyLine()` should be changed to private if not.

Reply:
Can you make a new PR to set `FunctionsMap.addFunction()` and `Helper.loadPolicyLine()` to private?

Reply:
ok I do another PR for this.

Instead of having a reference to CoreEnforcer in Model, why don't we stay with the modCount system? I find it cleaner because, as you said before, Model then remains a passive data object with no knowledge of CoreEnforcer. 

Moreover, it is not very elegant to manage this Model->CoreEnforcer "back reference":
- I have to call Model.setEnforcer at every places a Model can be set in CoreEnforcer
- in Model, I have to check that enforcer!=null because a model can be built without an enforcer

Reply:
I have pushed a simpler version (less changes). Now cache is invalidated when you call the following methods:

- Model.addDef
- ManagementEnforcer.(add|remove)(Named)GroupingPolicy
- ManagementEnforcer.addFunction

I put the method CoreEnforcer.resetExpressionEvaluator public so that a user can reset the cache if he uses some undocumented API.

I hope you will find this better.